File: java/leetcode-explore/algorithmInterview/array/IntersectionOfTwoArray.java
package algorithmInterview.array;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class IntersectionOfTwoArray {

    // 求两个数组的交集 可能有重复
    public int[] intersect(int[] nums1, int[] nums2) {
        if (nums1 == null || nums2 == null || nums1.length == 0 || nums2.length == 0) return new int[0];
        if (nums2.length > nums1.length) return intersect(nums2, nums1);
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums2) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        List<Integer> res = new ArrayList<>();
        for (int num : nums1) {
            if (map.containsKey(num) && map.get(num) > 0) {
                res.add(num);
                map.put(num, map.get(num) - 1);
            }
        }
        int[] resInts = new int[res.size()];
        for (int i = 0; i < res.size(); i++) {
            resInts[i] = res.get(i);
        }
        return resInts;
    }
}

File: go/binarySearch/intersectionOfTwoArray.go
package binarySearch

// 两个数组的交集
func intersection(nums1 []int, nums2 []int) []int {
	res := make([]int, 0)
	set := make(map[int]int)
	set2 := make(map[int]int)
	for _, num := range nums1 {
		set[num] = 1
	}
	for _, num := range nums2 {
		set2[num] = 1
	}

	for num := range set {
		if _, ok := set2[num]; ok {
			res = append(res, num)
		}
	}
	return res
}
